execution/function/function.go

```diff
@@ -0,0 +1,833 @@
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/value"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/thanos-community/promql-engine/execution/model"
+	"github.com/thanos-community/promql-engine/execution/parse"
+)
+
+// functionOperator returns []model.StepVector after processing input with desired function.
+type functionOperator struct {
+	funcExpr  *parser.Call
+	series    []labels.Labels
+	once      sync.Once
+	nextIndex int
+	nextFunc  func(ctx context.Context) ([]model.StepVector, error)
+	nextOps   []model.VectorOperator
+
+	call FunctionCall
+}
+
+func NewfunctionOperator(funcExpr *parser.Call, call FunctionCall, nextOps []model.VectorOperator) model.VectorOperator {
+	f := &functionOperator{
+		nextOps:   nextOps,
+		call:      call,
+		funcExpr:  funcExpr,
+		nextIndex: 0,
+	}
+
+	for i := range funcExpr.Args {
+		if funcExpr.Args[i].Type() == parser.ValueTypeMatrix || funcExpr.Args[i].Type() == parser.ValueTypeVector {
+			f.nextIndex = i
+		}
+	}
+
+	// Decide based on next type.
+	switch funcExpr.Args[f.nextIndex].Type() {
+	case parser.ValueTypeVector, parser.ValueTypeScalar:
+		f.nextFunc = f.instantVectorOrScalarNext
+	case parser.ValueTypeMatrix:
+		f.nextFunc = f.rangeVectorNext
+	default:
+		f.nextFunc = f.defaultNext
+	}
+
+	return f
+}
+
+func (o *functionOperator) Explain() (me string, next []model.VectorOperator) {
+	return fmt.Sprintf("[*functionOperator] %v(%v)", o.funcExpr.Func.Name, o.funcExpr.Args), o.nextOps
+}
+
+func (o *functionOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	return o.series, nil
+}
+
+func (o *functionOperator) GetPool() *model.VectorPool {
+	return o.nextOps[o.nextIndex].GetPool()
+}
+
+func (o *functionOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	// Process only non-variadic functions.
+	if o.funcExpr.Func.Variadic == 0 {
+		return o.nextFunc(ctx)
+	}
+
+	return nil, nil
+}
+
+// Process single/multi-arg instant vector and scalar input functions.
+func (o *functionOperator) instantVectorOrScalarNext(ctx context.Context) ([]model.StepVector, error) {
+	// Call next on instant/range vector input.
+	input, err := o.nextOps[o.nextIndex].Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	// Call next on all scalar inputs.
+	scalarInput := make([][]model.StepVector, len(o.nextOps)-1)
+	if len(scalarInput) != 0 {
+		s := 0
+		for i := range o.nextOps {
+			if i == o.nextIndex {
+				continue
+			}
+
+			si, err := o.nextOps[i].Next(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			scalarInput[s] = si
+			defer o.nextOps[i].GetPool().PutVectors(scalarInput[s])
+			s++
+		}
+	}
+
+	scalarPoints := make([]float64, len(scalarInput))
```

Comment:
Can be potentially be shared across `Next` invocations

Reply:
Have added slice to struct! 🙂